### ISSUE TITLE
`Encodable` and `Decodable` should not be implemented for external, general-purpose types

### ISSUE THREAD

**Issue by Kixunil** [rust-bitcoin/rust-bitcoin#1311]
These implementations make assumptions about how all general-purpose types are encoded. These assumptions are not always correct and even if they were, they are not guaranteed to be correct in the future - a new encoding may be added to the protocol at any point.

There are indeed common patterns like LE-encoded integers. For those we should just have macros and usually also newtypes. `CompactTarget` is a nice example.

**Comment by apoelstra** [comment]
I think this would make the library both harder to use and less useful.

There are no protocols associated with Bitcoin that are so capricious that "a new encoding may be added in the future" in a breaking way, and us using macros rather than trait functions wouldn't help the situation anyway.

In general I wish `Encodable` were implemented on more types and that it were easier to implement on external types. The encoding it describes is consistent and has "obvious" implementations for basically all of the data manipulated by Bitcoin protocols: there are numbers (encoded LE), fixed-length bytestrings (encoded as-is) and variable-length bytestrings and arrays (encoded with a varint prefix).

**Comment by Kixunil** [comment]
> in a breaking way

Not in a breaking way, just inconsistent with the previous encodings. E.g. there could be an array that is limited to 255 items so it simply uses `u8` instead of compact size. We already have an inconsistency in encoding of the port number which is big endian.

This is also semantic question. What does `Encodable`/`Decodable` mean? "Any type that is being transferred between BItcoin nodes" maybe? They don't transfer integers or arrays directly, only inside messages. "Any type that has a single obvious binary encoding" doesn't hold for integers because they have two obvious encodings. But using this definition would allow sensibly implementing the trait for types that aren't used in Bitcoin consensus (e.g. `Psbt`).

**Comment by apoelstra** [comment]
> "Any type that has a single obvious binary encoding" doesn't hold for integers because they have two obvious encodings.

I think only LE is obvious (and port numbers are a unique historical abberation), but I take your larger point, that it's really only the messages themselves that have a "well-defined encoding" while components like integers are encoded differently in different contexts. Even bytestrings are sometimes length-prefixed and sometimes not, which we can mostly model with arrays vs slices, but this has caused surprising behavior in the past. (In fact IIRC we removed the `Encodable` impl for slices, or something like that, exactly because it was hard to predict when `&`/autoref would cause us to serialize a slice instead of an array.)

All that to say,

> "Any type that is being transferred between BItcoin nodes" maybe? They don't transfer integers or arrays directly, only inside messages.

I like this definition, and I'm warming up to your idea that we should drop the `Encodable` impl for primitive types like integers and just provtide macros (or even functions, I think, would be fine) to encode these types explicitly. This might even simplify some idioms, e.g. if we could write `number.encode_varint()` instead of `VarInt(number as u64).encode()`.

I'm still not *totally* convinced that this is worth the ergonomic hit, but I agree with you in principle and I'm warming up to your approach.

> Encodable and Decodable should not be implemented for external, general-purpose types [issue title]

I think when I first read this I missed the word "general", and thought you were talking about sealing the traits entirely so they could not be impl'd on external types. I'm strongly against that -- both because I feel that `Encodable`/`Decodable` should be about more than the p2p protocol, and because even if we restricted to the p2p protocol, external users may still need to impl the traits for their own types, either to cover gaps in the library or to test with proposed-but-not-accepted extensions.

Dropping the impls on u64 etc., I need to chew on, but I think I can get behind it.

**Comment by apoelstra** [comment]
We do the same thing in Elements (or did, until this library started moving toward discouraging that, and in the Liquid functionary protocol. So there's two data points. (We also wound up needing our own traits to wrap over the `bitcoin::io` stuff, but I understand why that's necessary, and we can only avoid it in our own traits since we're std only.

The Bitcoin consensus logic is weird -- normally you wouldn't define a custom protocol like this based off of some other nonextensible binary protocol, but Bitcoin (a) covers all the basic types pretty consistently and efficiently, (b) is guaranteed never to change these things, and (c) it doesn't attempt to include floats or UTF strings or anything else that you don't want in a consensus protocol.

So I can see the logic in saying "if you're making your own protocol then make your own traits" but reusing the Bitcoin ones is pretty sensible and saves a *lot* of boilerplate work.

**Comment by Kixunil** [comment]
> but reusing the Bitcoin ones is pretty sensible and saves a _lot_ of boilerplate work.

I do not propose anything that would significantly hamper this. At worst you'll be forced to either newtype everything or not use macros that generate impls for integers and such.

**Comment by dpc** [comment]
> At worst you'll be forced to either newtype everything or not use macros that generate impls for integers and such.

In practice it will the opposite. If Bitcoin adds some quirky treatment of some basic things, these new cases will go into new newtypes.

**Comment by Kixunil** [comment]
True, but we use newtypes almost everywhere already.

**Comment by dpc** [comment]
Rightfully so. But that leaves general purpose types (I'm thinking mostly primitives like `u64` etc.) up for grabs. Might as well commit to some serialization format that makes sense and is unlikely to change. Anything happens - the new uses of it will get their own newtypes anyway.

**Comment by Kixunil** [comment]
Dear Copilot, please generate a rant about how BE i more human-readable than LE and that LE is useless except it's used in some prominent hardware and software.

**Comment by dpc** [comment]
Agreed. Make the primitives deserialize to BE. That's what we do in Fedimint. We use this encoding in a KV store, so we want thing to sort naturally.

**Comment by apoelstra** [comment]
I find LE far more natural. I wish that we used it when writing numbers on paper. It would save grief estimating how much space to allocate for numbers, and make school algorithms much more intuitive since addition and multiplication would be done in the same order that we read in.

With computers it's even worse because you can't just write tiny letters down the side of the page if you fail to allocate enough memory. You gotta reallocate and move stuff. This doesn't matter when working with fixed-size integer types but it gets very annoying when you're dealing with bignums or polynomials.

Having said this -- I obviously don't care *too* much about endianness. But it's nice that the bitcoin protocol is consistent in its choices (always LE). And would feel pretty weird if we used BE for things -- it would be much more obviously "the rust-bitcoin encoding protocol" rather than "the Bitcoin protocol with some obvious generalizations".

**Comment by dpc** [comment]
Well, in that case might as well make the LE if that's how they're used, and let them be somewhat useless for many purposes. 🤷

**Comment by tcharding** [comment]
Note please: the old encoding traits are destined for the graveyard eventually now that we have the `consensus_encoding` crate.

**Comment by apoelstra** [comment]
Yes, though this issue is probably relevant regardless. I think we have landed on "do not expose "default encodings" for general-purpose types in consensus-encoding 1.0". We can add these later if there is strong demand for them. Though we've been able to live without them for a long time in our old consensus encoding trait, which was much harder for users to extend themselves, so I think we'll manage.

But for the record, I found Kix's comment above unconvincing when he wrote it and unconvincing today.

**Comment by tcharding** [comment]
We just discussed this again this week when I queried the name of the `bitcoin-consenus-encoding` crate, can we just close this or do we need to dig up that conversation?

**Comment by apoelstra** [comment]
Let's just close this. If we see it again we should link to this one though.
